### PR TITLE
Simplified engine - BC breaking change

### LINKED ISSUE #795: Dataflow randomness problem with infinite data iterators
## 🐛 Bug description

```python
import torch
from ignite.engine import Engine, Events

def args_printing_wrapper(fn, prefix=""):    
    def wrapper(*args, **kwargs):
        print(prefix, *args, **kwargs)
        fn(*args, **kwargs)
    return wrapper

def random_train_data_generator():
    while True:
        yield torch.randint(0, 100, size=(4, ))

def random_val_data_generator():
    while True:
        yield torch.randint(0, 100, size=(4, )) + 100

def print_train_data(engine, batch):
    i = engine.state.iteration
    e = engine.state.epoch
    print("train", e, i, batch)

def print_val_data(engine, batch):
    i = engine.state.iteration
    print("val", i, batch)
    
trainer = Engine(print_train_data)
evaluator = Engine(print_val_data)

trainer._manual_seed = args_printing_wrapper(trainer._manual_seed, prefix="trainer")
evaluator._manual_seed = args_printing_wrapper(evaluator._manual_seed, prefix="evaluator")

@trainer.on(Events.EPOCH_COMPLETED)
def run_evaluation(_):
    evaluator.run(random_val_data_generator(), epoch_length=4)

trainer.run(random_train_data_generator(), max_epochs=4, epoch_length=6);
```
gives 
```
trainer 12 0
train 1 1 tensor([63,  3, 22, 77])
train 1 2 tensor([93, 86, 19, 83])
train 1 3 tensor([88, 20, 74, 93])
train 1 4 tensor([52, 25, 49, 41])
train 1 5 tensor([82, 87, 58, 18])
train 1 6 tensor([72,  9, 32, 64])
evaluator 12 0
val 1 tensor([163, 103, 122, 177])
val 2 tensor([193, 186, 119, 183])
val 3 tensor([188, 120, 174, 193])
val 4 tensor([152, 125, 149, 141])
train 2 7 tensor([82, 87, 58, 18])
train 2 8 tensor([72,  9, 32, 64])
train 2 9 tensor([77, 98, 57, 51])
train 2 10 tensor([58, 76, 39, 45])
train 2 11 tensor([82, 39, 27, 60])
train 2 12 tensor([58, 90,  2, 92])
evaluator 12 0
val 1 tensor([163, 103, 122, 177])
val 2 tensor([193, 186, 119, 183])
val 3 tensor([188, 120, 174, 193])
val 4 tensor([152, 125, 149, 141])
train 3 13 tensor([82, 87, 58, 18])
train 3 14 tensor([72,  9, 32, 64])
train 3 15 tensor([77, 98, 57, 51])
train 3 16 tensor([58, 76, 39, 45])
train 3 17 tensor([82, 39, 27, 60])
train 3 18 tensor([58, 90,  2, 92])
evaluator 12 0
val 1 tensor([163, 103, 122, 177])
val 2 tensor([193, 186, 119, 183])
val 3 tensor([188, 120, 174, 193])
val 4 tensor([152, 125, 149, 141])
train 4 19 tensor([82, 87, 58, 18])
train 4 20 tensor([72,  9, 32, 64])
train 4 21 tensor([77, 98, 57, 51])
train 4 22 tensor([58, 76, 39, 45])
train 4 23 tensor([82, 39, 27, 60])
train 4 24 tensor([58, 90,  2, 92])
evaluator 12 0
val 1 tensor([163, 103, 122, 177])
val 2 tensor([193, 186, 119, 183])
val 3 tensor([188, 120, 174, 193])
val 4 tensor([152, 125, 149, 141])
```
The problem is that `_manual_seed` is not called in this case and randomness of the data is configured by `evaluator`. Thus we have the same data all other epochs.

## Environment

 - PyTorch Version (e.g., 1.4): 1.4
 - Ignite Version (e.g., 0.3.0): 0.4.0.dev20200206
 - OS (e.g., Linux): Linux
 - How you installed Ignite (`conda`, `pip`, source): pip
 - Python version: 3.7
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by vfdev-5**
This PR is split into multiple PRs to ease the review:
- #938 (merged)
- #939 (merged)
- #9XX

--- 
Fixes #795 

Related to https://github.com/pytorch/ignite/pull/866

Description: 
Now, `Engine` runs without dataflow sync and thus has no issue with random states (e.g. #795)

Example save-resume on MNIST (https://github.com/pytorch/ignite/blob/19447d07bbe3e25a80bfa634add6f3fe75f1f52b/examples/mnist/mnist_save_resume_engine.py) :

```bash
rm -rf logs && python3 mnist_save_resume_engine.py --log_dir=logs --epochs 1 --deterministic && python3 mnist_save_resume_engine.py --log_dir=logs --epochs 1 --resume_from=logs/checkpoint_550.pt --deterministic
```
Gives the logs : 
Screen Shot 2020-04-18 at 03 05 31

We still provide an option to keep BC with v0.3.0:
- added module deterministic to be able sync dataflow:
```python

    import torch
    from torch.utils.data import DataLoader
    from ignite.engine import DeterministicEngine, Events
    from ignite.utils import manual_seed


    def random_train_data_loader(size):
        data = torch.arange(0, size)
        return DataLoader(data, batch_size=4, shuffle=True)


    def print_train_data(engine, batch):
        i = engine.state.iteration
        e = engine.state.epoch
        print("train", e, i, batch.tolist())

    trainer = DeterministicEngine(print_train_data)

    print("Original Run")
    manual_seed(56)
    trainer.run(random_train_data_loader(40), max_epochs=2, epoch_length=5)

    print("Resumed Run")
    # Resume from 2nd epoch
    trainer.load_state_dict({"epoch": 1, "epoch_length": 5, "max_epochs": 2, "rng_states": None})
    manual_seed(56)
    trainer.run(random_train_data_loader(40))

```

- Added new engine events (used with deterministic methods)
    - DATALOADER_STOP_ITERATION : engine's specific event triggered when dataloader has no more data to provide

    - TERMINATE_SINGLE_EPOCH : triggered when the run is about to end the current epoch,
      after receiving :meth:`~ignite.engine.Engine.terminate_epoch()` call.

    - TERMINATE : triggered when the run is about to end completely,
      after receiving :meth:`~ignite.engine.Engine.terminate()` call.

- Updated docs

TODO: 
* [x] Added tests to check data/grads/weights between original vs resumed training ([here](https://github.com/vfdev-5/ignite/blob/8cb2683a9af4a079fca893f76d43b320a7cfff0f/tests/ignite/engine/test_deterministic.py#L667))
* [ ] Example MNIST with gradient checking when resumed 
* [ ] Retrain example Cifar10
* [ ] Check https://github.com/pytorch/ignite/issues/818

Check list:
* [x] New tests are added (if a new feature is added)
* [x] New doc strings: description and/or example code are in RST format
* [x] Documentation is updated (if required)

cc @y0ast @sdesrozis

### CODE REVIEW

**Comment by vfdev-5** [comment]
Seems like this approach with `sync_dataflow=True` can lead to potential problems:
1) for different initial `set_seed(seed1)` and `set_seed(seed2)` dataflow will be randomly different before checkpointing and same after the checkpoint. 
=> This problem can be solved by introducing a random `seed_offset` : `checkpoint(i, to_save); set_seed(seed_offset + i)`
2) restart_data_loader() can lead to a unwanted data distribution. For example:
> dataloader is 250 steps long, user wants that many steps in epoch
checkpoint every 200 stepsAfter 200 steps, code saves checkpoint, dataloader resets
For next 50 steps, dataloader will not go through the leftover 1/5 of the data but instead restart

Data flow should be missing `state_dict` and `load_state_dict`-like methods.

**Comment by sdesrozis** [comment]
👍🏻

**Comment by vfdev-5** [comment]
Closed in favor of #940
